### PR TITLE
Add Plackett-Luce Distribution.

### LINKED ISSUE #38684: Add Plackett-Luce distribution
## 🚀 Feature
Add PlackettLuce and RelaxedPlackettLuce distributions. It is a simple distribution over permutations.

## Motivation

For optimization over categorical/binary variables (i.e. variational inference with discrete variational approximate posteriors, RL and etc) we already can use `RelaxedBernoulli` /`RelaxedOneHotCategorical` and `Bernoulli`/`OneHotCategorical` depending on if the method is relaxation based or not. 

Also, the same Gumbel-Softmax trick holds for the distribution over the set of permutations - Plackett-Luce distribution. You can see successful applications in https://arxiv.org/abs/1911.10036, https://arxiv.org/abs/1903.08850.

## Pitch

Tensorflow has tfp.distributions.PlackettLuce. I think it will be very cool to have it in the pytorch.

### PR DESCRIPTION
**Pull request by jeremysalwen**
Adds a new Distribution for permutations of N items, using the
Plackett-Luce distribution.  This can be useful for learning
orderings over sets of discrete items.

This implementation also supports batched distributions, where the
elements of the batch are different-sized permutations.

Fixes #38684

### CODE REVIEW

**Comment by jeremysalwen** [comment]
Can someone please review this?  There is no reason it should be closed.

**Comment by jeremysalwen** [comment]
@agolynski Are you supposed to be the reviewer for this PR?  Can you please reopen it?

Thanks,
Jeremy

**Comment by aldopareja** [comment]
please merge this
